scipy-stubs/signal/windows/_windows.pyi

```diff
@@ -29,36 +34,148 @@ __all__ = [
+
+_Window: TypeAlias = Literal[
+    "barthann",
+    "brthan",
+    "bth",
+    "bartlett",
+    "bart",
+    "brt",
+    "blackman",
+    "black",
+    "blk",
+    "blackmanharris",
+    "blackharr",
+    "bkh",
+    "bohman",
+    "bman",
+    "bmn",
+    "boxcar",
+    "box",
+    "ones",
+    "rect",
+    "rectangular",
+    "cosine",
+    "halfcosine",
+    "exponential",
+    "poisson",
+    "flattop",
+    "flat",
+    "flt",
+    "hamming",
+    "hamm",
+    "ham",
+    "hann",
+    "han",
+    "lanczos",
+    "sinc",
+    "nuttall",
+    "nutl",
+    "nut",
+    "parzen",
+    "parz",
+    "par",
+    "taylor",
+    "taylorwin",
+    "triangle",
+    "triang",
+    "tri",
+    "tukey",
+    "tuk",
+]
+_WindowNeedsParams: TypeAlias = Literal[
+    "chebwin",
+    "cheb",
+    "dpss",
+    "gaussian",
+    "gauss",
+    "gss",
+    "general cosine",
+    "general_cosine",
+    "general gaussian",
+    "general_gaussian",
+    "general gauss",
+    "general_gauss",
+    "ggs",
+    "general hamming",
+    "general_hamming",
+    "kaiser",
+    "ksr",
+    "kaiser bessel derived",
+    "kbd",
+]
+
+def general_cosine(M: int, a: Sequence[AnyReal], sym: bool = True) -> _Array_f8_1d: ...
+def boxcar(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def triang(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def parzen(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def bohman(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def blackman(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def nuttall(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def blackmanharris(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def flattop(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def bartlett(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def hann(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def tukey(M: int, alpha: AnyReal = 0.5, sym: bool = True) -> _Array_f8_1d: ...
+def barthann(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def general_hamming(M: int, alpha: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def hamming(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def kaiser(M: int, beta: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def kaiser_bessel_derived(M: int, beta: AnyReal, *, sym: bool = True) -> _Array_f8_1d: ...
+def gaussian(M: int, std: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def general_gaussian(M: int, p: AnyReal, sig: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def chebwin(M: int, at: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def cosine(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def exponential(M: int, center: AnyReal | None = None, tau: AnyReal = 1.0, sym: bool = True) -> _Array_f8_1d: ...
+def taylor(M: int, nbar: int = 4, sll: int = 30, norm: bool = True, sym: bool = True) -> _Array_f8_1d: ...
+def lanczos(M: int, *, sym: bool = True) -> _Array_f8_1d: ...
+
+#
+@overload
 def dpss(
-    M,
-    NW,
-    Kmax: Untyped | None = None,
+    *,
+    M: int,
+    NW: AnyReal,
+    Kmax: int = ...,
```

Comment:
Honestly I have clue what I am doing with the overloads. When you write an overload for arguments that have default values what values do you put in the overload case where the default value isn't a valid type?

Reply:
Overloads are kinda similar to pattern matching, but from the perspective of the caller.

So if you have e.g.

```python
def f(x=0, b=False):
    y = (2 << x) - 1
    return x if b else str(x)
```

For simplicity let's only assume builtin types for simplicity, i.e. `x: int` and `b: bool`. 
Because both `x` and `b` can be used positionally or with their keyword, there are four (actually 3, but I'll get to that) possible combinations:

1. `int`: `f()` or `f(x)` or `f(x, False)`
2. `int`: `f(b=False)` or `f(x, b=False)`
3. `str`: `f(b=True)` or `f(x, b=True)`
4. `str`: `f(x, True)`

So with this, the annotations will become

```pyi
from typing import Literal as L, overload

# 1.
@overload
def f(x: int = 0, b: L[False] = False) -> int: ...

# 2. (the `*` indicates that `b` is passed as keyword argument)
@overload
def f(x: int = 0, *, b: L[False]) -> int: ...

# 3.
@overload
def f(x: int = 0, *, b: L[True]) -> str: ... 

# 4. (note that `x` is required in this case, so it has no default)
@overload
def f(x: int, b: L[True]) -> str: ... 
```

Note that the second overload is actually redundant: The first one will already match `f(b=False)` or `f(x, b=False)`, and both overloads have the same return type. So that leaves us with

```pyi
@overload
def f(x: int = 0, b: L[False] = False) -> int: ...
@overload
def f(x: int = 0, *, b: L[True]) -> str: ... 
@overload
def f(x: int, b: L[True]) -> str: ... 
```

In this case, the order doesn't matter, because none of the overloads overlap, so you could equivalently write this as e.g.

```pyi
@overload
def f(x: int, b: L[True]) -> str: ... 
@overload
def f(x: int = 0, *, b: L[True]) -> str: ... 
@overload
def f(x: int = 0, b: L[False] = False) -> int: ...
```

So the order is just a matter of personal preference in this case (although it might slightly affect the performance of type-checker) 🤷🏻.